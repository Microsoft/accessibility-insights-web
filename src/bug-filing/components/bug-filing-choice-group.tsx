// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ChoiceGroup, IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';
import * as React from 'react';

import { UserConfigMessageCreator } from '../../common/message-creators/user-config-message-creator';
import { NamedSFC } from '../../common/react/named-sfc';
import { BugFilingService } from '../types/bug-filing-service';

export type BugFilingChoiceGroupProps = {
    deps: BugFilingChoiceGroupDeps;
    selectedBugFilingService: BugFilingService;
    bugFilingServices: BugFilingService[];
};

export type BugFilingChoiceGroupDeps = {
    userConfigMessageCreator: UserConfigMessageCreator;
};

export const BugFilingChoiceGroup = NamedSFC<BugFilingChoiceGroupProps>('BugFilingChoiceGroup', props => {
    const { userConfigMessageCreator } = props.deps;

    const getOptions: () => IChoiceGroupOption[] = () => {
        return props.bugFilingServices.map(service => {
            return {
                key: service.key,
                text: service.displayName,
            };
        });
    };

    const onChange = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: IChoiceGroupOption) => {
        userConfigMessageCreator.setBugService(option.key);
    };

    return (
        <ChoiceGroup
            className={'bug-filing-choice-group'}
            onChange={onChange}
            options={getOptions()}
            selectedKey={props.selectedBugFilingService.key}
        />
    );
});
