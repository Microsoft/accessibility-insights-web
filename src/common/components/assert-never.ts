// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// This function helps with exhaustiveness checking in switch statements
// If used in the default case, TS will throw an error unless all cases are handled.
export const assertNever = (x: never): never => {
    throw Error(
        'This error should never be thrown when strict null checks are applied to this project',
    );
};
