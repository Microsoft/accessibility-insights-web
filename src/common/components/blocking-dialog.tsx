// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DefaultButton, PrimaryButton } from 'office-ui-fabric-react/lib/Button';
import { Dialog, DialogFooter, DialogType, IDialogProps } from 'office-ui-fabric-react/lib/Dialog';
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';

export type BlockingDialogProps = IDialogProps;

export const BlockingDialog = NamedSFC<BlockingDialogProps>('BlockingDialog', props => {
    return (
        // As of writing, NVDA has an issue where it doesn't announce new elements with role=alertdialog,
        // which Office Fabric applies implicitly whenever isBlocking: true is set. As a workaround, our
        // "blocking" dialog sets isBlocking: false (which means the dialog will use role=dialog instead of
        // role=alertdialog, which NVDA announces more reliably), and emulates being "blocking" by making
        // sure we don't set any onDismiss handlers and overriding showCloseButton to false.
        <Dialog
            {...props}
            onDismiss={undefined}
            dialogContentProps={{
                ...props.dialogContentProps,
                showCloseButton: false,
            }}
            modalProps={{
                ...props.modalProps,
                // Be warned that changing this value can change which child element is focused by default
                // (we saw this in the telemetry-permission-dialog); be sure to test this behavior specifically
                // if you ever change this back to true.
                isBlocking: false,
                onDismiss: undefined,
            }}
        />
    );
});
