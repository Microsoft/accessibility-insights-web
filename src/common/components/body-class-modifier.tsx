// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DocumentManipulator } from 'common/document-manipulator';
import * as React from 'react';

export type BodyClassModifierProps = {
    classNames: string[];
    documentManipulator: DocumentManipulator;
};

export class BodyClassModifier extends React.Component<BodyClassModifierProps> {
    public render(): JSX.Element {
        return null;
    }

    private updateClassNamesFrom(prevClassNames: string[]): void {
        const { classNames, documentManipulator } = this.props;

        const addedClassNames = classNames.filter(n => !prevClassNames.includes(n));
        const removedClassNames = prevClassNames.filter(n => !classNames.includes(n));

        // Note: may include class names not managed by us; leave them as-is
        let allBodyClassNames = documentManipulator.bodyClassNames;
        allBodyClassNames = allBodyClassNames.filter(n => !removedClassNames.includes(n));
        allBodyClassNames.push(...addedClassNames);

        documentManipulator.bodyClassNames = allBodyClassNames;
    }

    public componentDidMount(): void {
        this.updateClassNamesFrom([]);
    }

    public componentDidUpdate(prevProps: BodyClassModifierProps): void {
        this.updateClassNamesFrom(prevProps.classNames);
    }
}
