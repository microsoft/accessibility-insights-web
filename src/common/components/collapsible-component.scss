// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

@import '../../common/styles/common.scss';

span.collapsible-title {
    padding-left: 0.5vw;
}

.collapsible {
    padding: 0;
    cursor: pointer;
    width: 100%;
    border: none;
    outline: none;
    line-height: 0.25px;
    display: flex;
    justify-content: flex-start;
    align-items: center;

    h2 {
        font-weight: $normal-title-font-weight;
        font-size: 17px;
    }
}

.collapsible-content {
    overflow: hidden;
    padding-left: calc(0.5vw + 13px); // 13px is Chevron width.
}
