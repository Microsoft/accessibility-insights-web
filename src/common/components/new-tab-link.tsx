// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BaseComponent } from '@uifabric/utilities';
import { ILinkProps, Link } from 'office-ui-fabric-react/lib/Link';
import * as React from 'react';

export class NewTabLink extends BaseComponent<ILinkProps> {
    public render(): JSX.Element {
        return <Link className="insights-link" {...this.props} target="_blank" />;
    }
}
