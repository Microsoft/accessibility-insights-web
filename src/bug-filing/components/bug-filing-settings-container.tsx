// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';
import { BugServiceProperties } from '../../common/types/store-data/user-configuration-store';
import { SettingsDeps } from '../../DetailsView/components/settings-panel/settings/settings-props';
import { BugFilingServiceProvider } from '../bug-filing-service-provider';
import { BugFilingService } from '../types/bug-filing-service';
import { BugFilingChoiceGroup, BugFilingChoiceGroupDeps } from './bug-filing-choice-group';

export interface BugFilingSettingsContainerProps {
    deps: BugFilingSettingsContainerDeps;
    selectedBugFilingService: BugFilingService;
    selectedBugFilingServiceData: BugServiceProperties;
}

export type BugFilingSettingsContainerDeps = {
    bugServiceProvider: BugFilingServiceProvider;
} & BugFilingChoiceGroupDeps &
    SettingsDeps;

export const BugFilingSettingsContainer = NamedSFC<BugFilingSettingsContainerProps>('BugFilingSettingsContainer', props => {
    const { deps, selectedBugFilingService, selectedBugFilingServiceData } = props;
    const SettingsForm = selectedBugFilingService.renderSettingsForm;
    const bugFilingServices = deps.bugServiceProvider.allVisible();

    return (
        <>
            <BugFilingChoiceGroup deps={deps} selectedBugFilingService={selectedBugFilingService} bugFilingServices={bugFilingServices} />
            <SettingsForm deps={deps} settings={selectedBugFilingServiceData} />
        </>
    );
});
