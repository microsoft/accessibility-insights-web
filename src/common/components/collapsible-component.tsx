// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ActionButton } from '@fluentui/react';
import { css } from '@fluentui/utilities';
import * as styles from 'common/components/collapsible-component.scss';
import * as React from 'react';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        this.setState(prevState => ({ showContent: !prevState.showContent }));
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, styles.collapsibleContent)}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName)}>
                <ActionButton
                    className={styles.collapsible}
                    iconProps={{ iconName: iconName }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    {this.props.header}
                </ActionButton>
                {content}
            </div>
        );
    }
}
