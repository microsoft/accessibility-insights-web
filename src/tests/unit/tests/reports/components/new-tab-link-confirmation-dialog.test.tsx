// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NewTabLink } from 'common/components/new-tab-link';
import { shallow } from 'enzyme';
import { isFunction } from 'lodash';
import * as React from 'react';
import {
    ConfirmType,
    NewTabLinkWithConfirmationDialog,
} from 'reports/components/new-tab-link-confirmation-dialog';
import { IMock, It, Mock, MockBehavior, Times } from 'typemoq';

describe('NewTabLinkWithConfirmationDialog', () => {
    it('generates unique id for each link', () => {
        const testSubject1 = shallow(<NewTabLinkWithConfirmationDialog />);
        const testSubject2 = shallow(<NewTabLinkWithConfirmationDialog />);

        expect(testSubject1.find(NewTabLink).prop('id')).not.toEqual(
            testSubject2.find(NewTabLink).prop('id'),
        );
    });

    it('contains NewTabLink with forwarded props', () => {
        const props = {
            href: 'test-href',
            title: 'test-title',
        };
        const child = <div>test-child</div>;

        const testSubject = shallow(
            <NewTabLinkWithConfirmationDialog {...props}>{child}</NewTabLinkWithConfirmationDialog>,
        );

        const testSubjectNewTabLink = testSubject.find(NewTabLink);

        const { id, ...testableProps } = testSubjectNewTabLink.props();

        expect(testableProps).toMatchSnapshot();
    });

    describe('onClickHandler', () => {
        const targetPageLinkMock = Mock.ofType<HTMLElement>();

        type GetElementById = typeof Document.prototype.getElementById;

        let getElementByIdMock: IMock<GetElementById>;
        let confirmMock: IMock<ConfirmType>;

        let originalGetElementById: GetElementById;
        let originalConfirm: ConfirmType;

        beforeEach(() => {
            originalGetElementById = document.getElementById;
            originalConfirm = window.confirm;

            getElementByIdMock = Mock.ofType<GetElementById>(undefined, MockBehavior.Strict);
            confirmMock = Mock.ofType<ConfirmType>(undefined, MockBehavior.Strict);

            document.getElementById = getElementByIdMock.object;
            window.confirm = confirmMock.object;
        });

        afterEach(() => {
            document.getElementById = originalGetElementById;
            window.confirm = originalConfirm;
        });

        it('does not use IE-incompatible arrow function syntax', () => {
            const testSubject = shallow(<NewTabLinkWithConfirmationDialog />);
            const generatedScript = testSubject.find('script').render().html();

            expect(generatedScript).not.toMatch(/=>/);
        });

        it('is added to the link', () => {
            targetPageLinkMock
                .setup(link => link.addEventListener('click', It.is(isFunction)))
                .verifiable(Times.once());

            const testSubject = shallow(<NewTabLinkWithConfirmationDialog />);

            const linkId = testSubject.find(NewTabLink).prop('id');
            getElementByIdMock
                .setup(handler => handler(linkId))
                .returns(() => targetPageLinkMock.object)
                .verifiable(Times.once());

            const generatedScript = testSubject.find('script').render().html();

            // tslint:disable-next-line: no-eval
            eval(generatedScript);

            targetPageLinkMock.verifyAll();
            getElementByIdMock.verifyAll();
        });

        it('calls confirm function with expected text', () => {
            const eventMock = Mock.ofType<Event>();
            eventMock.setup(event => event.preventDefault());

            getElementByIdMock
                .setup(handler => handler(It.isAnyString()))
                .returns(() => targetPageLinkMock.object);

            confirmMock
                .setup(handler => handler(It.isAnyString()))
                .callback(message => expect(message).toMatchSnapshot())
                .verifiable(Times.once());

            let clickListener: Function;
            targetPageLinkMock
                .setup(link => link.addEventListener('click', It.is(isFunction)))
                .callback((_, listener) => (clickListener = listener));

            const testSubject = shallow(<NewTabLinkWithConfirmationDialog />);

            const generatedScript = testSubject.find('script').render().html();

            // tslint:disable-next-line: no-eval
            eval(generatedScript);

            clickListener(eventMock.object);

            confirmMock.verifyAll();
        });

        it('allows link to be followed when OK is selected on confirmation dialog', () => {
            const eventMock = Mock.ofType<Event>();
            eventMock.setup(event => event.preventDefault()).verifiable(Times.never());

            getElementByIdMock
                .setup(handler => handler(It.isAnyString()))
                .returns(() => targetPageLinkMock.object);

            confirmMock.setup(handler => handler(It.isAnyString())).returns(() => true);

            let clickListener: Function;
            targetPageLinkMock
                .setup(link => link.addEventListener('click', It.is(isFunction)))
                .callback((_, listener) => (clickListener = listener));

            const testSubject = shallow(<NewTabLinkWithConfirmationDialog />);

            const generatedScript = testSubject.find('script').render().html();

            // tslint:disable-next-line: no-eval
            eval(generatedScript);

            clickListener(eventMock.object);

            eventMock.verifyAll();
        });

        it('prevents link from being followed when Cancel is selected on confirmation dialog', () => {
            const eventMock = Mock.ofType<Event>();
            eventMock.setup(event => event.preventDefault()).verifiable(Times.once());

            getElementByIdMock
                .setup(handler => handler(It.isAnyString()))
                .returns(() => targetPageLinkMock.object);

            confirmMock.setup(handler => handler(It.isAnyString())).returns(() => false);

            let clickListener: Function;
            targetPageLinkMock
                .setup(link => link.addEventListener('click', It.is(isFunction)))
                .callback((_, listener) => (clickListener = listener));

            const testSubject = shallow(<NewTabLinkWithConfirmationDialog />);

            const generatedScript = testSubject.find('script').render().html();

            // tslint:disable-next-line: no-eval
            eval(generatedScript);

            clickListener(eventMock.object);

            eventMock.verifyAll();
        });
    });
});
