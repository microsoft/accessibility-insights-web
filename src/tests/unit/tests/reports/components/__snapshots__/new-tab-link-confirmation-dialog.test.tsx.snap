// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTabLinkWithConfirmationDialog contains NewTabLink with forwarded props 1`] = `
{
  "children": <div>
    test-child
  </div>,
  "href": "test-href",
  "title": "test-title",
}
`;

exports[`NewTabLinkWithConfirmationDialog onClickHandler calls confirm function with expected text 1`] = `
"Are you sure you want to navigate away from the Accessibility Insights report?
This link will open the target page in a new tab.

Press OK to continue or Cancel to stay on the current page."
`;